src/ILCompiler.Compiler/src/Compiler/InteropStubManager.cs

```diff
@@ -101,32 +103,169 @@ internal TypeDesc GetInlineArrayType(InlineArrayCandidate candidate)
+            public MethodDesc ClosedDelegateMarshallingThunk;
+            public MethodDesc DelegateCreationThunk;
         }
 
         internal IEnumerable<DelegateMarshallingThunks> GetDelegateMarshallingThunks()
         {
             foreach (var delegateType in _delegateMarshalingTypes)
             {
-                var openStub = InteropStateManager.GetOpenStaticDelegateMarshallingThunk(delegateType);
-                var closedStub = InteropStateManager.GetClosedDelegateMarshallingThunk(delegateType);
-                var delegateCreationStub = InteropStateManager.GetForwardDelegateCreationThunk(delegateType);
                 yield return 
                     new DelegateMarshallingThunks()
                     {
                         DelegateType = delegateType,
-                        OpenStaticDelegateMarshallingThunk = openStub,
-                        ClosedDelegateMarshallingThunk = closedStub,
-                        DelegateCreationThunk = delegateCreationStub
+                        OpenStaticDelegateMarshallingThunk = InteropStateManager.GetOpenStaticDelegateMarshallingThunk(delegateType),
+                        ClosedDelegateMarshallingThunk = InteropStateManager.GetClosedDelegateMarshallingThunk(delegateType),
+                        DelegateCreationThunk = InteropStateManager.GetForwardDelegateCreationThunk(delegateType)
                     };
             }
         }
 
-        internal HashSet<NativeStructType> GetStructMarshallingTypes()
+        internal struct StructMarshallingThunks
         {
-            return _structMarshallingTypes;
+            public TypeDesc StructType;
+            public NativeStructType NativeStructType;
+            public MethodDesc MarshallingThunk;
+            public MethodDesc UnmarshallingThunk;
+            public MethodDesc CleanupThunk;
+        }
+
+        internal IEnumerable<StructMarshallingThunks> GetStructMarshallingTypes()
+        {
+                foreach (var structType in _structMarshallingTypes)
+            {
+                yield return
+                    new StructMarshallingThunks()
+                    {
+                        StructType = structType,
+                        NativeStructType = InteropStateManager.GetStructMarshallingNativeType(structType),
+                        MarshallingThunk = InteropStateManager.GetStructMarshallingManagedToNativeThunk(structType),
+                        UnmarshallingThunk = InteropStateManager.GetStructMarshallingNativeToManagedThunk(structType),
+                        CleanupThunk = InteropStateManager.GetStructMarshallingCleanupThunk(structType)
+                    };
+            }
+        }
+
+        public void AddDependeciesDueToPInvoke(ref DependencyList dependencies, NodeFactory factory, MethodDesc method)
+        {
+            // On Project N, the compiler doesn't generate the interop code on the fly
+            if (method.Context.Target.Abi != TargetAbi.ProjectN)
```

Comment:
Nit: Might be nicer to keep the `if` check on the `CodeBasedBlah` side because the fact that Project X differs is not interop's problem - it's more of a `CodeBasedBlah`'s problem.